### PR TITLE
Add exec function that returns the stdout and stderr of the container

### PR DESCRIPTION
**Pull request by fabiogouw**
This PR address the issue #390 , I'm taking the approach that creates a new method so no usage is broken. Please let me know if it's ok or if it needs any kind of improvement. Thank you.

### CODE REVIEW

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 1]
```diff
@@ -137,6 +137,38 @@ await this.Docker.Exec.StartContainerExecAsync(created.ID, ct)
...
+      {
+        Cmd = command,
+        AttachStderr = true,
+        AttachStdout = true,
+      };
+
+      var created = await this.Docker.Exec.ExecCreateContainerAsync(id, execCreateParameters, ct)
+        .ConfigureAwait(false);
+
+      var stream = await this.Docker.Exec.StartAndAttachContainerExecAsync(created.ID, false, ct);
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs]: `.ConfigureAwait(false);`

File: src/DotNet.Testcontainers/Clients/ExecResult.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+namespace DotNet.Testcontainers.Clients
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/ExecResult.cs]: Move the file to `src\DotNet.Testcontainers\Containers\`, namespace `DotNet.Testcontainers.Containers`.

File: src/DotNet.Testcontainers/Clients/ExecResult.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+{
+  public class ExecResult
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/ExecResult.cs]: Please use constructor initialization.

```cs
public readonly struct ExecResult
{
  public ExecResult(string stdOut, string stdErr, long exitCode)
  {
    this.StdOut = stdOut;
    this.StdErr = stdErr;
    this.ExitCode = exitCode;
  }

  public string StdOut { get; }

  public string StdErr { get; }

  public long ExitCode { get; }
}
```

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 4]
```diff
@@ -137,6 +137,38 @@ await this.Docker.Exec.StartContainerExecAsync(created.ID, ct)
...
+      var results = new ExecResult();
+
+      for (ContainerExecInspectResponse response; (response = await this.Docker.Exec.InspectContainerExecAsync(created.ID, ct)
+        .ConfigureAwait(false)) != null;)
+      {
+        if (!response.Running)
+        {
+          (results.StdOut, results.StdErr) = await stream.ReadOutputToEndAsync(ct);
+          results.ExitCode = response.ExitCode;
+          break;
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs]: `return new ExecResult(stdOut, stdErr, response.ExitCode);` see `ExecResult` suggestions.

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 5]
```diff
@@ -137,6 +137,38 @@ await this.Docker.Exec.StartContainerExecAsync(created.ID, ct)
...
+        }
+      }
+
+      return results;
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs]: `return new ExecResult(string.Empty, string.Empty, long.MinValue)`.

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 6]
```diff
@@ -137,6 +137,38 @@ await this.Docker.Exec.StartContainerExecAsync(created.ID, ct)
...
+      var stream = await this.Docker.Exec.StartAndAttachContainerExecAsync(created.ID, false, ct);
+
+      var results = new ExecResult();
+
+      for (ContainerExecInspectResponse response; (response = await this.Docker.Exec.InspectContainerExecAsync(created.ID, ct)
+        .ConfigureAwait(false)) != null;)
+      {
+        if (!response.Running)
+        {
+          (results.StdOut, results.StdErr) = await stream.ReadOutputToEndAsync(ct);
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs]: `.ConfigureAwait(false);`

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
PR looks good! Thanks. I suggest only a few small changes. If you don't mind, I think we should use the existing `ExecAsync` method. Even it is a breaking change, I think it's better to stay close to the Java implementation. Version 1.6 has breaking changes anyway (renaming namespaces).

Thanks again.

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs (hunk 1 above)
Comment (fabiogouw) [reply src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs]: Done!

**Comment by fabiogouw** [comment]
Thanks for your considerations, @HofmeisterAn , I made all the changes. However, the pipeline now hangs in the Tests stage, I'll take a look at the root cause.

**Comment by HofmeisterAn** [comment]
> Thanks for your considerations, @HofmeisterAn , I made all the changes. However, the pipeline now hangs in the Tests stage, I'll take a look at the root cause.

It's because the wait strategies never return, see:

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/8b3860c0bb27d8f82f747fdba07870ef33a18bad/src/DotNet.Testcontainers/Configurations/WaitStrategies/UntilUnixCommandIsCompleted.cs#L29

We need to double check the `ExecAsync` calles. I'll take a look at it tomorrow, but I'm pretty sure there are not many left.

**Comment by fabiogouw** [comment]
> > Thanks for your considerations, @HofmeisterAn , I made all the changes. However, the pipeline now hangs in the Tests stage, I'll take a look at the root cause.
> 
> It's because the wait strategies never return, see:
> 
> https://github.com/HofmeisterAn/dotnet-testcontainers/blob/8b3860c0bb27d8f82f747fdba07870ef33a18bad/src/DotNet.Testcontainers/Configurations/WaitStrategies/UntilUnixCommandIsCompleted.cs#L29
> 
> We need to double check the `ExecAsync` calles. I'll take a look at it tomorrow, but I'm pretty sure there are not many left.

Great, that was the problem! Thanks

**Review by HofmeisterAn — APPROVED** [review]

**Pull request merged**